### ISSUE TITLE
Build fails `ApiCompat error : 0 : Unable to resolve assembly 'Assembly(Name=System.Private.CoreLib`

### ISSUE THREAD

**Issue by nietras** [dotnet/corefx#25750]
I have tried building corefx `master` branch for some time on Windows. It has yet to succeed for me. I've tried on multiple windows machines. Followed the build instructions for windows. Yet it never works. Tried different commits and lately the one below with Visual Studio 15.5 installed. Still fails.

Almost everytime I want to do a contribution I hit these things, spent countless hours building and failing, trying to figure out what is wrong, but fail. It is a bit frustrating 🙄 

I even tried a release branch but it still failed.

What am I doing wrong?

commit: 749194e8d9813d1116520530042c261474341ce8
[init-tools.log](https://github.com/dotnet/corefx/files/1534914/init-tools.log)
[msbuild.log](https://github.com/dotnet/corefx/files/1534897/msbuild.log)
```
EXEC : ApiCompat error : 0 : Unable to resolve assembly 'Assembly(Name=System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)' referenced by the implementation assembly 'Ass
embly(Name=mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)'. [D:\oss\corefx\src\shims\ApiCompat.proj]
EXEC : ApiCompat error : 0 : Unable to resolve assembly 'Assembly(Name=System.Private.Uri, Version=4.0.5.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)' referenced by the implementation assembly 'Assembl
y(Name=System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)'. [D:\oss\corefx\src\shims\ApiCompat.proj]
EXEC : ApiCompat error : 0 : Unable to resolve assembly 'Assembly(Name=System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)' referenced by the implementation assembly 'Ass
embly(Name=System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)'. [D:\oss\corefx\src\shims\ApiCompat.proj]
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.CriticalHandleMinusOneIsInvalid' does not exist in the implementation but it does exist in the contract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.CriticalHandleZeroOrMinusOneIsInvalid' does not exist in the implementation but it does exist in the con
tract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.SafeFileHandle' does not exist in the implementation but it does exist in the contract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.SafeHandleMinusOneIsInvalid' does not exist in the implementation but it does exist in the contract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.SafeHandleZeroOrMinusOneIsInvalid' does not exist in the implementation but it does exist in the contrac
t.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'Microsoft.Win32.SafeHandles.SafeWaitHandle' does not exist in the implementation but it does exist in the contract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'System.AccessViolationException' does not exist in the implementation but it does exist in the contract.
D:\oss\corefx\src\shims\ApiCompat.proj(76,5): error : TypesMustExist : Type 'System.Action' does not exist in the implementation but it does exist in the contract.
```

**Comment by karelz** [comment]
@weshaggard any idea what might cause it? I haven't seen similar report so far :(

**Comment by weshaggard** [comment]
I think the other error in your log is the root cause:
```
         D:\oss\corefx\Tools\Packaging.targets(1068,5): error : Files D:\oss\corefx\bin\AnyOS.AnyCPU.Debug\mscorlib\netcoreapp\mscorlib.dll and D:\oss\corefx\bin\obj\shims\netcoreapp\facades\mscorlib.dll have the same TargetPath Microsoft.Private.CoreFx.NETCoreApp/ref/netcoreapp2.1/mscorlib.dll. [D:\oss\corefx\pkg\Microsoft.Private.CoreFx.NETCoreApp\Microsoft.Private.CoreFx.NETCoreApp.pkgproj]
         D:\oss\corefx\Tools\Packaging.targets(1068,5): error : Files D:\oss\corefx\bin\obj\shims\netcoreapp\facades\System.Data.dll and D:\oss\corefx\bin\AnyOS.AnyCPU.Debug\System.Data\netcoreapp\System.Data.dll have the same TargetPath Microsoft.Private.CoreFx.NETCoreApp/ref/netcoreapp2.1/System.Data.dll. [D:\oss\corefx\pkg\Microsoft.Private.CoreFx.NETCoreApp\Microsoft.Private.CoreFx.NETCoreApp.pkgproj]
         D:\oss\corefx\Tools\Packaging.targets(1068,5): error : Files D:\oss\corefx\bin\obj\shims\netcoreapp\facades\System.dll and D:\oss\corefx\bin\AnyOS.AnyCPU.Debug\System\netcoreapp\System.dll have the same TargetPath Microsoft.Private.CoreFx.NETCoreApp/ref/netcoreapp2.1/System.dll. [D:\oss\corefx\pkg\Microsoft.Private.CoreFx.NETCoreApp\Microsoft.Private.CoreFx.NETCoreApp.pkgproj]

```

I've seen this once before because of a bug in msbuild (see https://github.com/dotnet/corefx/pull/24645/commits/d0e3851776a64cea638ae2c53f67a3b25def5756) which caused both sets of facades to be bin-placed. However I thought my fix (assuming you have it on your machine) should have taken care of this. I cannot see any obvious reason this would still fail.

@nietras can you please trying doing a `clean.cmd --all` and then a `build.cmd` to see if that helps, I assume it won't given you have tried on multiple machines, however please give it a try and then I will help figure out the issue if it still exists.

**Comment by nietras** [comment]
> please trying doing a clean.cmd --all and then a build.cmd to see if that helps

Running `clean.cmd --all` did indeed solve this, had run `clean.cmd`, but `--all` will be my default in the future 😄 Thanks.

I did have another issue though. Opening `System.Memory.sln` in VS 15.5 and building would give the following error for the `System.Memory.Performance.Tests` project (and lots of warnings):
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0006	Metadata file 'E:\oss\corefx\bin\AnyOS.AnyCPU.Debug\PerfRunner\netstandard\PerfRunner.exe' could not be found	System.Memory.Performance.Tests	E:\oss\corefx\src\System.Memory\tests\Performance\CSC	1	N/A
```
Running `build-tests.cmd` solved that, since apparently the `PerfRunner.exe` was missing otherwise.

To recap, I had to run:
```
clean.cmd --all
build.cmd
build-tests.cmd
```
To get to a good state. Perhaps this could be put into the build instructions as a general recommendation to run for people not doing corefx work everyday. As far as I can tell the build instructions do not explicitly say you should run `build-tests.cmd` before opening a solution in VS.

Perhaps it should be considered to add a `build-all.cmd` or `all.cmd` that builds everything and perhaps even cleans first, to have a single command that should work no matter what and gets the repo into a useable state.

Now I have the issue though that running `System.Memory.Tests` via F5 in VS for example will run `OuterLoop` tests too, which is pretty annoying given how long it takes. How can I run inner loop tests only from within VS?

Also running individual tests via `CTRL+R + T` does not work either, but that is a known issue currently I think.

**Comment by nietras** [comment]
On the other machine I had issues, the `clean.cmd --all` worked too, but for some reasons the `build-tests.cmd` fails with:
```
  === TEST EXECUTION SUMMARY ===
D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Linq.Tests  Total: 3938, Errors: 0, Failed: 4, Skipped: 6, Time: 4.627s [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
  Finished running tests.  End time=10:51:13.76, Exit code = 1
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Linq.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/netcoreapp-Windows_NT-
Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Linq.Tests' please check D:\oss\corefx\bin/tests/System.Linq.Tests/netcoreapp-Windows_NT-Debug-x64/te
stResults.xml for details! [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
  Finished:    System.Net.Primitives.UnitTests.Tests

  === TEST EXECUTION SUMMARY ===
D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Net.Primitives.UnitTests.Tests  Total: 69, Errors: 0, Failed: 1, Skipped: 0, Time: 4.431s [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\
System.Net.Primitives.UnitTests.Tests.csproj]
  Finished running tests.  End time=10:51:14.82, Exit code = 1
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Net.Primitives.UnitTests.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/ne
tcoreapp-Windows_NT-Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\System.Net.Primitives.UnitTests.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Net.Primitives.UnitTests.Tests' please check D:\oss\corefx\bin/tests/System.Net.Primitives.UnitTests.
Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\System.Net.Primitives.UnitTests.Tests.csproj]
  Finished:    System.Drawing.Common.Tests

  === TEST EXECUTION SUMMARY ===
D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Drawing.Common.Tests  Total: 4506, Errors: 0, Failed: 1, Skipped: 0, Time: 6.436s [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Com
mon.Tests.csproj]
  Finished running tests.  End time=10:51:16.79, Exit code = 1
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Drawing.Common.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/netcoreapp-W
indows_NT-Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Common.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Drawing.Common.Tests' please check D:\oss\corefx\bin/tests/System.Drawing.Common.Tests/netcoreapp-Win
dows_NT-Debug-x64/testResults.xml for details! [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Common.Tests.csproj]
D:\oss\corefx\dir.traversal.targets(77,5): error : (No message specified) [D:\oss\corefx\src\tests.builds]

Build FAILED.

D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Linq.Tests  Total: 3938, Errors: 0, Failed: 4, Skipped: 6, Time: 4.627s [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Linq.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/netcoreapp-Windows_NT-
Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Net.Primitives.UnitTests.Tests  Total: 69, Errors: 0, Failed: 1, Skipped: 0, Time: 4.431s [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\
System.Net.Primitives.UnitTests.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Net.Primitives.UnitTests.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/ne
tcoreapp-Windows_NT-Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\System.Net.Primitives.UnitTests.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(484,5): warning :    System.Drawing.Common.Tests  Total: 4506, Errors: 0, Failed: 1, Skipped: 0, Time: 6.436s [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Com
mon.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(484,5): warning MSB3073: The command "D:\oss\corefx\bin/tests/System.Drawing.Common.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd D:\oss\corefx\bin/testhost/netcoreapp-W
indows_NT-Debug-x64/" exited with code 1. [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Common.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Linq.Tests' please check D:\oss\corefx\bin/tests/System.Linq.Tests/netcoreapp-Windows_NT-Debug-x64/te
stResults.xml for details! [D:\oss\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Net.Primitives.UnitTests.Tests' please check D:\oss\corefx\bin/tests/System.Net.Primitives.UnitTests.
Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [D:\oss\corefx\src\System.Net.Primitives\tests\UnitTests\System.Net.Primitives.UnitTests.Tests.csproj]
D:\oss\corefx\Tools\tests.targets(492,5): error : One or more tests failed while running tests from 'System.Drawing.Common.Tests' please check D:\oss\corefx\bin/tests/System.Drawing.Common.Tests/netcoreapp-Win
dows_NT-Debug-x64/testResults.xml for details! [D:\oss\corefx\src\System.Drawing.Common\tests\System.Drawing.Common.Tests.csproj]
D:\oss\corefx\dir.traversal.targets(77,5): error : (No message specified) [D:\oss\corefx\src\tests.builds]
    6 Warning(s)
    4 Error(s)

Time Elapsed 00:00:59.54
Command execution failed with exit code 1.
```

**Comment by weshaggard** [comment]
@nietras I'm glad that fixed it for you. Out of curiosity did read the "Common full clean build and test run" section in https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md? It outlines exactly the steps you mention. If there is something you think would help in the docs please submit a PR to update them.

As for our Outerloop vs Innerloop test runs in VS I wouldn't expect the outerloop tests to be run be default. Can you look at the command line options in Debug Properties tab for the test project? You can customize the xunit commandline options that are used in there.

>Also running individual tests via CTRL+R + T does not work either, but that is a known issue currently I think.

Yes I think that is a limitation currently because we haven't done the work to hook in our test runner into VS. 

As for the test errors on your other machine I think the test results just need to be examined to see why they are failing.

I think we have you mostly unblocked now so I'm closing this issue but please feel free to ask more questions if necessary. Thanks for your interest in contributing to our project.
